### PR TITLE
[dy] Support reading authentication settings from AWS secrets manager

### PR DESCRIPTION
**Pull request by dy46**
# Description

This PR adds support for reading settings from AWS secrets manager. Only some of the existing settings have been moved to using the new `Settings` class. Most server settings that are not secrets are still being directly read from environment variables. 

In order to set the settings backend, the user will need to add a section to the project metadata. Here is an example to set the setting backend to AWS secrets manager:

```
settings_backend:
  backend_type: aws_secrets_manager
  prefix: default_repo/settings/
```

This PR also adds an observer to the project metadata file so that if there are any updates to that file, the settings backend can be updated.

# How Has This Been Tested?

- [x] Tested with AWS secrets manager backend and reading settings
- [x] Tested existing auth methods
  - [x]  GHE
  - [x]  Gitlab
  - [x]  Bitbucket
  - [x]  Git
  - [ ]  Okta
  - [x]  Google
  - [x]  AD
  - [x]  OIDC

# Checklist
- [x] The PR is tagged with proper labels (bug, enhancement, feature, documentation)
- [x] I have performed a self-review of my own code
- [ ] I have added unit tests that prove my fix is effective or that my feature works
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] If new documentation has been added, relative paths have been added to the appropriate section of `docs/mint.json`

cc:

### CODE REVIEW

File: mage_ai/data_preparation/repo_manager.py [hunk 1]
```diff
@@ -300,6 +304,13 @@ def set_project_uuid_from_metadata() -> None:
             project_uuid = config.get('project_uuid')
 
 
+def update_settings_on_metadata_change() -> None:
+    global project_uuid
```
Comment (wangxiaoyou1993) [inline mage_ai/data_preparation/repo_manager.py:308]: where is the project_uuid used?

File: mage_ai/settings/backends.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        if self.prefix:
+            key = f'{self.prefix}{key}'
+        try:
+            secret_response = self.client.get_secret_value(
+                SecretId=key,
+            )
+        except ClientError as error:
+            if error.response['Error']['Code'] == 'ResourceNotFoundException':
+                return None
+            raise
```
Comment (wangxiaoyou1993) [inline mage_ai/settings/backends.py:80]: what are other possible errors? do we not fall back to env variables for those errors?

File: mage_ai/settings/backends.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    """
+    backend_type = BackendType.AWS_SECRETS_MANAGER
+
+    def __init__(self, **kwargs) -> None:
+        import boto3
+
+        self.client = boto3.client('secretsmanager')
+        self.prefix = kwargs.get('prefix', '')
+
+    def _get(self, key: str, **kwargs) -> Optional[str]:
```
Comment (wangxiaoyou1993) [inline mage_ai/settings/backends.py:69]: can we also allow user to configure whether to use AWS secret cache?

Comment (dy46) [reply mage_ai/settings/backends.py:69]: good idea, i'll add that

File: mage_ai/data_preparation/repo_manager.py (hunk 1 above)
Comment (dy46) [reply mage_ai/data_preparation/repo_manager.py:308]: oops i'll remove that

File: mage_ai/settings/backends.py (hunk 2 above)
Comment (dy46) [reply mage_ai/settings/backends.py:80]: good point. i'll log an error and return None

**Comment by dy46** [comment]
@wangxiaoyou1993 PTAL

**Comment by wangxiaoyou1993** [comment]
Have you tested the with those different authorization methods?

**Comment by wangxiaoyou1993** [comment]
can you also update the doc?

**Review by wangxiaoyou1993 — APPROVED** [review]

**Pull request merged**